compiler/src/dotty/tools/dotc/typer/RefChecks.scala

```diff
@@ -782,7 +782,50 @@ object RefChecks {
         report.error(problem(), clazz.srcPos)
       }
 
+      // Checks base types of the current clazz against each other
+      //
+      // In particular, it checks that there are no two base classes with
+      // different type instantiations.
+      def validateBaseTypes(): Unit = {
+        val tpe = clazz.thisType // in Scala 2 this was clazz.tpe
+        val seenParents = mutable.HashSet[Type]()
+        val baseClasses: List[ClassSymbol] = clazz.info.baseClasses
+
+        // tracks types that we have seen for a particular base class in baseClasses
+        val seenTypes = MutableSymbolMap[List[Type]](baseClasses.size)
+
+        // validate all base types of a class in reverse linear order.
+        def register(tp: Type): Unit = {
+          val baseClass = tp.classSymbol
+          if (baseClasses contains baseClass) {
+            val alreadySeen = seenTypes.getOrElse(baseClass, Nil)
+            if (!alreadySeen.exists(_ <:< tp))
+              seenTypes.update(baseClass, tp :: alreadySeen.filterNot(tp <:< _))
+          }
+          val remaining = tp.parents filterNot seenParents
+          seenParents ++= remaining
+          remaining foreach register
+        }
+        register(tpe)
+
+        seenTypes.iterator.foreach {
+          case (cls, Nil) =>
+            assert(false) // this case should not be reachable
```

Comment:
Taking advantage of https://github.com/lampepfl/dotty/blob/ce684de6ce4ce77c17750c675343d65a5ef24137/compiler/src/dotty/tools/package.scala#L54:
```suggestion
            unreachable()
```